### PR TITLE
DownloadTableVC functionality and Code cleanup

### PR DESCRIPTION
**Pull request by kreonjr**

### CODE REVIEW

File: Podverse/DownloadsTableViewController.swift [hunk 1]
```diff
@@ -9,27 +9,150 @@
...
-
-        // Do any additional setup after loading the view.
     }
-
-    override func didReceiveMemoryWarning() {
-        super.didReceiveMemoryWarning()
-        // Dispose of any resources that can be recreated.
+    
+    override func viewWillAppear(_ animated: Bool) {
+        PVDownloader.shared.delegate = self
```
Comment (kreonjr) [inline Podverse/DownloadsTableViewController.swift]: This line needs to go into viewDidLoad

File: Podverse/DownloadsTableViewController.swift [hunk 2]
```diff
@@ -9,27 +9,150 @@
...
 
-    /*
-    // MARK: - Navigation
+}
 
-    // In a storyboard-based application, you will often want to do a little preparation before navigation
+extension DownloadsTableViewController:UITableViewDelegate, UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 92
```
Comment (kreonjr) [inline Podverse/DownloadsTableViewController.swift]: Lets make this a constant :)

File: Podverse/DownloadsTableViewController.swift [hunk 3]
```diff
@@ -9,27 +9,150 @@
...
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! DownloadTableViewCell
+        
+        let episode = DownloadingEpisodeList.shared.downloadingEpisodes[indexPath.row]
+        
+        DispatchQueue.global().async {
+            Podcast.retrievePodcastUIImage(downloadingEpisode: episode) { (podcastImage) -> Void in
+                DispatchQueue.main.async {
```
Comment (kreonjr) [inline Podverse/DownloadsTableViewController.swift:52]: Double dispatch

File: Podverse/DownloadsTableViewController.swift [hunk 4]
```diff
@@ -9,27 +9,150 @@
...
+        
+        tableView.deselectRow(at: indexPath, animated: false)
+
+    }
+    
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
-        // Get the new view controller using segue.destinationViewController.
-        // Pass the selected object to the new view controller.
+        if segue.identifier == "To Now Playing" {
+            let mediaPlayerViewController = segue.destination as! MediaPlayerViewController
```
Comment (kreonjr) [inline Podverse/DownloadsTableViewController.swift]: if let here for safety

File: Podverse/DownloadsTableViewController.swift [hunk 5]
```diff
@@ -9,27 +9,150 @@
...
+        }
     }
-    */
+    
+}
 
+extension DownloadsTableViewController:PVDownloaderDelegate {
+    func downloadFinished(episode: DownloadingEpisode) {
+        if let indexPath = indexPathOfDownload(episode: episode) {
+            if let cell = tableView.cellForRow(at: indexPath) as? DownloadTableViewCell {
```
Comment (kreonjr) [inline Podverse/DownloadsTableViewController.swift]: one chained unwrap

File: Podverse/EpisodesTableViewController.swift [hunk 6]
```diff
@@ -84,59 +71,16 @@ class EpisodesTableViewController: PVViewController, UITableViewDataSource, UITa
...
-//    
+
     override func viewDidLoad() {
         super.viewDidLoad()
-        self.loadData()
+        loadData()
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
         PVDownloader.shared.delegate = self
```
Comment (kreonjr) [inline Podverse/EpisodesTableViewController.swift:81]: Move to viewDidLoad

File: Podverse/EpisodesTableViewController.swift [hunk 7]
```diff
@@ -215,66 +151,78 @@ class EpisodesTableViewController: PVViewController, UITableViewDataSource, UITa
...
-//                    self.pvMediaPlayer.loadEpisodeDownloadedMediaFileOrStream(selectedEp.objectID, paused: false)
-//                    self.segueToNowPlaying()
-//                }))
-//            } else {
-//                if selectedEp.taskIdentifier != nil {
-//                    episodeActions.addAction(UIAlertAction(title: "Downloading Episode", style: .Default, handler: nil))
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "To Now Playing" {
+            let mediaPlayerViewController = segue.destination as! MediaPlayerViewController
```
Comment (kreonjr) [inline Podverse/EpisodesTableViewController.swift]: unwrap for safety

File: Podverse/PVViewController.swift [hunk 8]
```diff
@@ -9,10 +9,93 @@
...
+                playPause.setTitleColor(UIColor.gray, for: .normal)
+                
+                if (pvMediaPlayer.avPlayer.rate > 0) {
+                    playPause.setImage(UIImage(named: "Pause"), for: .normal)
+                } else {
+                    playPause.setImage(UIImage(named: "Play"), for: .normal)
+                }
+                
+                // TODO: why does this playOrPause selector fail? how do we fix it?
+                let playPauseGesture = UITapGestureRecognizer(target: self, action: #selector (pvMediaPlayer.playOrPause))
```
Comment (kreonjr) [inline Podverse/PVViewController.swift]: you need a selector within this class that itself calls playOrPause. I don't think recognizers are smart enough to go look at other files for their selectors

File: Podverse/PVViewController.swift [hunk 9]
```diff
@@ -9,10 +9,93 @@
...
+                playPause.addGestureRecognizer(playPauseGesture)
+                nowPlayingBar.addSubview(playPause)
+                
+                let bottomDivider:UIView = UIView(frame: CGRect(x: 0, y: 50, width: self.view.bounds.size.width, height: 1))
+                bottomDivider.backgroundColor = UIColor.lightGray
+                nowPlayingBar.addSubview(bottomDivider)
+                
+                self.view!.addSubview(nowPlayingBar)
```
Comment (kreonjr) [inline Podverse/PVViewController.swift]: !

File: Podverse/EpisodesTableViewController.swift [hunk 10]
```diff
@@ -169,13 +111,9 @@ class EpisodesTableViewController: PVViewController, UITableViewDataSource, UITa
                 cell.summary?.text = summary.removeHTMLFromString()
             }
 
-            let totalClips = String(episode.clips.count)
+            let totalClips = String(123)
```
Comment (mitchdowney) [inline Podverse/EpisodesTableViewController.swift:114]: no

File: Podverse/EpisodesTableViewController.swift (hunk 6 above)
Comment (kreonjr) [reply Podverse/EpisodesTableViewController.swift:81]: will revisit

**Pull request merged**